extensions/amp-analytics/0.1/variables.js

```diff
@@ -0,0 +1,276 @@
+import {fromClass} from '../../../src/service';
+import {isArray} from '../../../src/types';
+import {cryptoFor} from '../../../src/crypto';
+
+const TAG = 'Analytics.Variables';
+
+/** @typedef {function(...?):(?|!Promise<?>)} */
+let VariableFilterDef;
+
+/**
+ * @param {!string} str
+ * @param {Number} s
+ * @param {Number} l
+ * @return {string}
+ */
+function substrFilter(str, s, l) {
+  const start = Number(s);
+  let length = str.length;
+  user().assertNumber(start,
+      'Start index ' + start + 'in substr filter should be a number');
+  if (l) {
+    length = user().assertNumber(Number(l),
+        'Length ' + length + ' in substr filter should be a number');
+  }
+
+  return str.substr(start, length);
+}
+
+/**
+ * @param {*} value
+ * @param {*} defaultValue
+ * @return {*}
+ */
+function defaultFilter(value, defaultValue) {
+  return value || user().assertString(defaultValue);
+}
+
+
+export class VariableService {
+  /**
+   * @param {!Window} window
+   */
+  constructor(window) {
+
+    /** @private {!Window} */
+    this.win_ = window;
+
+    /** @private {!Object<string, VariableFilterDef>} */
+    this.filters_ = Object.create(null);
+
+      /** @private {./crypto-impl.Crypto} */
+    this.crypto_ = null;
+
+    this.register_('default', defaultFilter);
+    this.register_('substr', substrFilter);
+    this.register_('trim', value => user().assertString(value).trim());
+    this.register_('json', value => JSON.stringify(value));
+    this.register_('toLowerCase', value =>
+        user().assertString(value).toLowerCase());
+    this.register_('toUpperCase', value =>
+        user().assertString(value).toUpperCase());
+    this.register_('not', value => String(!value));
+    this.register_('base64', value => btoa(String(value)));
+    this.register_('hash', this.hashFilter_.bind(this));
+    this.register_('if', (value, thenValue, elseValue) =>
+        Promise.resolve(value ? thenValue : elseValue));
+
+    cryptoFor(this.win_).then(crypto => {
+      this.crypto_ = crypto;
+    });
+  }
+
+  /**
+   * @param {string} name
+   * @param {VariableFilterDef} handler
+   */
+  register_(name, handler) {
+    dev().assert(!this.filters_[name], 'Filter "' + name
+        + '" already registered.');
+    dev().assert(handler, 'Handler for filter ' + name + ' is invalid.');
+    this.filters_[name] = handler;
+  }
+
+  /**
+   * @param {string} filterStr
+   * @return {!Object<string, *>}
+   */
+  parseFilter_(filterStr) {
+    if (!filterStr) {
+      return {};
+    }
+
+    // The parsing for filters breaks when `:` is used as something other than
+    // the argument separator. A full-fledged parser would be needed to fix
+    // this.
+    const tokens = filterStr.split(':');
+    user().assert(tokens[0], 'Filter ' + name + ' is invalid.');
+    const fn = this.filters_[tokens[0]];
+    if (!fn) {
+      user().error(TAG, 'Invalid filter name: ' + tokens[0]);
+      return {};
+    }
+    return {fn, args: tokens.slice(1)};
```

Comment:
This is not addressed. Not sure if this should be done at this level or inside the functions.